pkg/modprovider/state.go

```diff
@@ -200,46 +227,105 @@ func (h *moduleStateHandler) Diff(
 		return nil, err
 	}
-	moduleInputDiff := !oldInputs["moduleInputs"].DeepEquals(newInputs["moduleInputs"])
 
-	if !newState.Equal(oldState) || moduleInputDiff {
-		changes = pulumirpc.DiffResponse_DIFF_SOME
+	if !plan.HasChanges() {
+		// Since Create/Update will not be called after this, to avoid hangs we need to make sure the unchanged
+		// TF state is propagated to the rest of the system. Currently this is done approximately by running a
+		// no-op apply for its side-effects.  It would be more efficient to do mod.planStore.SetState if
+		// we had the capacity to hydrate the State from the serialized form in oldState.
+		mod.apply(ctx, logger, tf)
+
+		return &pulumirpc.DiffResponse{Changes: pulumirpc.DiffResponse_DIFF_NONE}, nil
 	}
 
-	return &pulumirpc.DiffResponse{Changes: changes}, nil
+	return &pulumirpc.DiffResponse{Changes: pulumirpc.DiffResponse_DIFF_SOME}, nil
 }
 
-// Create exposes empty old state and returns the new state.
+// Create runs tofu apply.
 func (h *moduleStateHandler) Create(
-	_ context.Context,
+	ctx context.Context,
 	req *pulumirpc.CreateRequest,
 ) (*pulumirpc.CreateResponse, error) {
-	oldState := moduleState{}
-	modUrn := h.mustParseModURN(req.Properties)
-	h.oldState.Put(modUrn, oldState)
-	newState := h.newState.Await(modUrn)
-	props := newState.Marshal()
+	logger := newResourceLogger(h.hc, resource.URN(req.GetUrn()))
+
+	var args moduleStateResourceArgs
+	if err := args.Unmarshal(req.Properties); err != nil {
+		return nil, err
+	}
+
+	mod := h.mod(args.ModuleURN)
+
+	// Diff does not run before Create, but it is paramount to have a TF plan. Therefore run the plan here.
+	wd := tfsandbox.ModuleInstanceWorkdir(args.ModuleURN)
+	tf, err := tfsandbox.NewTofu(ctx, wd, h.auxProviderServer)
+	if err != nil {
+		return nil, fmt.Errorf("Sandbox construction failed: %w", err)
+	}
+	h.tofuCache[args.ModuleURN] = tf
+
+	plan, err := mod.plan(ctx, logger, tf, args.ModuleInputs, moduleState{})
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Preview {
+		return &pulumirpc.CreateResponse{
+			Id:         moduleStateResourceID,
+			Properties: (&moduleState{moduleOutputs: plan.Outputs()}).Marshal(),
+		}, nil
+	}
+
+	newState, _, applyErr := mod.apply(ctx, logger, tf)
+	if applyErr != nil {
+		return nil, applyErr // should this be handled better for partial apply?
+	}
+
 	return &pulumirpc.CreateResponse{
 		Id:         moduleStateResourceID,
-		Properties: props,
+		Properties: newState.Marshal(),
 	}, nil
 }
 
-// Update simply returns the new state.
+// Create runs tofu apply.
 func (h *moduleStateHandler) Update(
-	_ context.Context,
+	ctx context.Context,
 	req *pulumirpc.UpdateRequest,
 ) (*pulumirpc.UpdateResponse, error) {
-	newState := h.newState.Await(h.mustParseModURN(req.News))
+	logger := newResourceLogger(h.hc, resource.URN(req.GetUrn()))
+
+	var args moduleStateResourceArgs
+	if err := args.Unmarshal(req.News); err != nil {
+		return nil, err
+	}
+
+	mod := h.mod(args.ModuleURN)
+
+	if req.Preview {
```

Comment:
This could add outputs to the preview on the ModuleState resource; not sure that's super helpful though as presumably this is also available as component outputs. 

Also, a quick bug is that plan.Outputs() exposes unwanted internal secret-tracking properties.